text/0000-javascript-module-api.md

```diff
@@ -0,0 +1,1058 @@
+- Start Date: 2016-11-05
+- RFC PR: https://github.com/emberjs/rfcs/pull/176
+- Ember Issue: (leave this empty)
+
+# Summary
+
+Make Ember feel less overwhelming to new users, and make Ember applications
+start faster, by replacing the `Ember` global with a first-class system for
+importing just the parts of the framework you need.
+
+# Motivation
+
+ECMAScript 2015 (also known as ES2015 or ES6) introduced a syntax for importing
+and exporting values from modules. Ember aggressively adopted modules, and if
+you've used Ember before, you're probably familiar with this syntax:
+
+```js
+import Ember from "ember";
+import Analytics from "../mixins/analytics";
+
+export default Ember.Component.extend(Analytics, {
+  // ...
+});
+```
+
+One thing to notice is that the entire Ember framework is imported as a single
+package. Rather than importing `Component` directly, for example, you import
+`Ember` and subclass `Ember.Component`. (And this example still works even if
+you forget the import, because we also create a global variable called
+`Ember`.)
+
+Using Ember via a monolithic package or global object is not ideal for several
+reasons:
+
+* It's overwhelming for learners. There's a giant list of classes and functions
+  with no hints about how they're related. The API documentation reflects this.
+* Experienced developers who don't want all of Ember's features feel like
+  they're adding unnecessary and inescapable bloat to their application.
+* The `Ember` object must be built at boot time, requiring that we ship the
+  entire framework to the browser. This has two major costs:
+  1. Increased download time, particularly noticeable on slower connections.
+  2. Increased parsing/evaluation cost, which still must be paid even when
+     assets are cached. On some browsers/devices, this can far exceed the cost of the
+     download itself.
+* Running multiple versions of Ember on the same page is error-prone.
+
+Defining a public API for importing parts of Ember via JavaScript modules helps
+us lay the groundwork for solving all of these problems.
+
+#### Multiple Apps on a Page
```

Comment:
We're looking to _reduce_ the frequency of this type of behavior in browser-side code on the CLI side. Further, because of the way we're using Broccoli, you get even-more pernicious behavior (`full outer join`) with clobbering at the intersection (`overwrite: true`) based upon which is "last" thing we see (level + DAG topsort + package.json ordering).

I'd rather us not present this as a goal/motivation for this RFC.

---

Short version of goals:
- Shipped-to-browser code should follow the Highlander rule. (`--flat`)
- Server-side build related things should (probably) allow for npm-style module resolution.